crates/wasmtime/src/runtime/vm/traphandlers/backtrace.rs

```diff
@@ -176,6 +236,100 @@ impl Backtrace {
+        unwind: &dyn Unwind,
+        chain: VMStackChain,
+        pc: usize,
+        fp: usize,
+        trampoline_fp: usize,
+        mut f: impl FnMut(Frame) -> ControlFlow<()>,
+    ) -> ControlFlow<()> {
+        use crate::runtime::vm::stack_switching::{VMContRef, VMStackLimits};
+
+        // Handle the stack that is currently running (which may be a
+        // continuation or the initial stack).
+        Self::trace_through_wasm(unwind, pc, fp, trampoline_fp, &mut f)?;
+
+        // Note that the rest of this function has no effect if `chain` is
+        // `Some(VMStackChain::InitialStack(_))` (i.e., there is only one stack to
+        // trace through: the initial stack)
+
+        assert_ne!(chain, VMStackChain::Absent);
+        let stack_limits_vec: Vec<*mut VMStackLimits> =
+            chain.clone().into_stack_limits_iter().collect();
+        let continuations_vec: Vec<*mut VMContRef> =
+            chain.clone().into_continuation_iter().collect();
+
+        // The VMStackLimits of the currently running stack (whether that's a
+        // continuation or the initial stack) contains undefined data, the
+        // information about that stack is saved in the Store's
+        // `VMRuntimeLimits` and handled at the top of this function
+        // already. That's why we ignore `stack_limits_vec[0]`.
+        //
+        // Note that a continuation stack's control context stores
+        // information about how to resume execution *in its parent*. Thus,
+        // we combine the information from continuations_vec[i] with
+        // stack_limits_vec[i + 1] below to get information about a
+        // particular stack.
+        //
+        // There must be exactly one more `VMStackLimits` object than there
+        // are continuations, due to the initial stack having one, too.
+        assert_eq!(stack_limits_vec.len(), continuations_vec.len() + 1);
+
+        for i in 0..continuations_vec.len() {
+            let (continuation, parent_continuation, parent_limits) = unsafe {
+                // The continuation whose control context we want to
+                // access, to get information about how to continue
+                // execution in its parent.
+                let continuation = &*continuations_vec[i];
+
+                // The stack limits describing the parent of `continuation`.
+                let parent_limits = &*stack_limits_vec[i + 1];
+
+                // The parent of `continuation`, if the parent is itself a
+                // continuation. Otherwise, if `continuation` is the last
+                // continuation (i.e., its parent is the initial stack), this is
+                // None.
+                let parent_continuation = if i + 1 < continuations_vec.len() {
+                    Some(&*continuations_vec[i + 1])
+                } else {
+                    None
+                };
+                (continuation, parent_continuation, parent_limits)
+            };
```

Comment:
Can this be rewritten to use [slice::chunks](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.chunks) and `zip`? And also narrow the `unsafe` to just the relevant blocks (because I had a hard time figuring out why it was needed at a glance). I think this would make the code a bit safer and easier to read. Something like this (although preserving comments where appropriate):

```suggestion
        for (conts, parent_limits) in continuations_vec.chunks(2).zip(stack_limits_vec.iter().skip(1)) {
            let continuation = conts[0];
            let continuation = unsafe { &*continuation };
            
            let parent_limits = unsafe { &*parent_limits };
            
            let parent_continuation = conts.get(1).map(|p| unsafe { &*p });
            
            // ...
```